README.md

Slidemapper
===========

A jQuery plugin for creating slideshows on a leaflet map.  Will render a slideshow-carousel linked to a map inside selected elements, extending to the full width of the container.  The map will follow along as the user arrows/clicks through the geotagged slides.

Check out [the demo here](http://cav.is/slidemapper/example/index.html "Demo").  Unless you're busy.  Cause I totally understand.


![Alt text](http://cav.is/img/slidemapper-example.png "Slidemapper Demo")

Requirements
------------

### jQuery

Requires [jQuery](http://jquery.com/ "jQuery") version **>= 1.7.0**

### Leaflet

Requires [Leaflet](http://leaflet.cloudmade.com/ "Leaflet") version **>= 0.4**.  Since this is still in development, I'd suggest getting it directly from [their github repo](https://github.com/CloudMade/Leaflet "github").

### Leafpile

If you're going to use the `leafPile: true` clustering option, you should also include [leafpile.js](http://github.com/cavis/leafpile "Leafpile").  Otherwise, don't worry about it.


Getting Started
------------

Include either the minified or standard versions of both slidemapper.js and slidemapper.css.  (See example/index.html for an example).

### In your `<head>` section (next to other CSS):

    <link href="path/to/leaflet.css" rel="stylesheet">
    <link href="path/to/slidemapper.min.css" rel="stylesheet">

### At the bottom of the `<body>` section (next to other JS)

    <script src="path/to/jquery-1.7.2.min.js"></script>
    <script src="path/to/leaflet.js"></script>
    <script src="path/to/slidemapper.min.js"></script>


Usage
------------

Slidemapper is just a jQuery plugin, so just call the function on a selector to create a slideshow.  This is the equivalent of calling the `init` method on the element.

    $(document).ready(function() {
        var options = {};
        $mySlideMap = $('#slideshow-container').slideMapper(options);
    });

Then, you can call other Slidemapper methods by passing the string method name as the first argument, like so:

    $mySlideMap.slideMapper('add', someSlideConfig);
    $mySlideMap.slideMapper('count');
    $('#slideshow-container').slideMapper('removeAll');


Options
------------

These options can be passed in when you initialize the slideMapper

`slides` - An optional array of slide configurations to add as soon as the slideshow is created.

`mapType` - String indicating the type of map tiles to use.  Valid options are `cloudmade`, `stamen-toner`, `stamen-terrain`, `stamen-watercolor`, `mapquest`, and `mapquest-aerial`.  If you choose `cloudmade`, you must also provide an API key. - __(default: `mapquest`)__

`apiKey` - Optional string API key for use with `cloudmade` tiles.  Get one [here](http://developers.cloudmade.com/ "Cloudmade").

`center` - The starting/default center lat/lng for the map - __(default: `[40.423, -98.7372]`)__

`zoom` - The starting/default zoom level for the map - __(default: 4)__

`minZoom` - The minimum zoom level the map should allow - __(default: 2)__

`maxZoom` - The maximum zoom level the map should allow - __(default: 10)__

`enableKeyEvents` - Enable key events (left-right arrow keys) - __(default: true)__

`closePopupOnClick` - Allow popups to close when the map is clicked - __(default: false)__

`mapPosition` - The location of the map relative to the slideshow; `top` or `bottom` - __(default: `bottom`)__

`mapHeight` - The height in pixels of the map - __(default: 400)__

`slideHeight` - The height in pixels of the slideshow element.  If `autoHeight` is true, this becomes the min-height of the slideshow element. - __(default: `220`)__

`autoHeight` - Automatically resize the height of the slideshow element to match the height of the current slide - __(default: false)__

`leafPile` - Enable clustering of markers on the map using an `L.Leafpile`.  May set to `true`, or set to a config object to be passed to the constructor of the `L.Leafpile`. - __(default: false)__

`animateSpeed` - The speed in milliseconds of the slideshow animations - __(default: 200)__

`controlType` - Which type of slideshow controls to display; `sides` or `top` - __(default: `sides`)__


Methods
------------

Manipulate the slides in the show

`add`__(config)__ - add a slide to the end of the slideshow

`insert`__(index, config)__ - insert a slide into the slideshow

`get`__()__ - get the currently displayed slide

`get`__(index)__ - get a slide at an index

`count`__()__ - get the total number of slides

`shuffle`__(moveCurrentToIndex)__ - move the currently displayed slide to a new index

`shuffle`__(indexFrom, indexTo)__ - move a slide to a new index

`remove`__()__ - remove the currently displayed slide from the slideshow

`remove`__(index)__ - remove a specific slide from the slideshow

`removeAll`__()__ - remove all slides from this slideshow

Change the currently displayed slide

`move`__(index, animate)__ - move to a slide at an index, optionally animating the transition

`next`__()__ - move to the next slide

`prev`__()__ - move to the previous slide

Enable and disable events

`keyEvents`__(turnOn)__ - enable or disable key events

`mapEvents`__(turnOn)__ - enable or disable map events

`freeze`__(makeFrozen)__ - freeze the show, preventing the user from changing slides or scrolling/zooming the map


Events
------------

`move` - fired when the slide changes

    $mySlideMap = $('#slideshow-container').slideMapper();
    $mySlideMap.on('move', function(e, slide, index) {
        // e     - the event object
        // slide - the slide we're moving to
        // index - the index of the slide we're moving to
        if (slide.index == 4) {
            return false; // prevent the slide from changing
        }
    });


Issues and Contributing
-----------------------

Report any bugs or feature-requests via the issue tracker.  Or send me a fax.  And if you'd like to contribute, send me a note!  Thanks.


License
------------

Leafpile is free software, and may be redistributed under the MIT-LICENSE.

Thanks for listening!


CHANGELOG.md

SlideMapper Versions
====================

0.1.0
------------

* Initial "good enough for jazz" version.
* Support for multiple slide configurations
* Clustering support (via Leafpile)

0.1.1
------------

* Updated to latest leaflet lib
* Updated to latest leapfile lib
* fixed bug where clicking the same marker twice triggers slide

0.1.2
------------

* Ability to use a custom image icon


example/data.js
EXAMPLEDATA = [
  // intro marker
  {
    icon: 'other.png',
    marker: [42.516846, -70.898499],
    center: [40.423, -98.7372],
    html: '<table style="margin:0 40px; padding:10px"><tr>' +
            '<td><img src="http://placehold.it/300x180&text=Map+Stuff"/></td>' +
            '<td style="padding-left:10px">' +
              '<h1>SlideMapper FTW!</h1>' +
              '<p>This is a demo of the different sorts of slides you can setup using slidemapper.</p>' +
            '</td>' +
          '</tr></table>',
    popup: 'So it begins!'
  },

  // marker with everything
  {
    marker: [42.295006, -85.622467],
    center: [42.295006, -85.622467],
    zoom: 4,
    html: '<table style="margin:0 40px; padding:10px"><tr>' +
            '<td style="padding-left:10px">' +
              '<h3>Configure all options</h3>' +
              '<p>Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras justo odio, dapibus ac facilisis in, egestas eget quam.</p>' +
            '</td>' +
            '<td><img src="http://placehold.it/300x180&text=Some+Image"/></td>' +
          '</tr></table>',
    popup: 'zoom level 4'
  },

  // marker without center/zoom set ... will inherit from the previous slide
  {
    marker: [47.683819, -122.301224],
    html: '<div style="margin:0 40px; padding:20px 10px">' +
            '<div>' +
              '<h2>A whole bunch of overflowing text</h2>' +
              '<img src="http://placehold.it/350x100&text=Another+Image" style="float:right; padding:10px"/>' +
              '<p>Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Nulla vitae elit libero, a pharetra augue. Integer posuere erat a ante venenatis <a href="#">dapibus posuere</a> velit aliquet.</p><p>Cras justo odio, dapibus ac facilisis in, egestas eget quam. Sed posuere consectetur est at lobortis. Aenean lacinia bibendum nulla sed consectetur. Fusce dapibus, tellus ac cursus commodo, tortor mauris condimentum nibh, ut fermentum massa justo sit amet risus. Aenean eu leo quam. Pellentesque ornare sem lacinia quam venenatis vestibulum.</p>' +
            '</div>' +
          '</div>',
    popup: 'no zoom set... stays the same'
  },

  // full-image slide
  {
    marker: [33.053093, -117.268547],
    zoom: 7,
    html: '<img src="http://placehold.it/1000x600&text=Full+Size+Image" style="width:100%;height:100%;margin-bottom:-3px"/>' +
          '<span style="position:absolute;bottom:0;right:0;background:#333;color:#fff;padding:4px 10px">This is a fullsize image... it won\'t scroll at all</span>',
    popup: 'change to zoom level 7'
  },

  // slide without a marker
  {
    center: [40.423, -98.7372],
    zoom: 7,
    html: '<div style="margin:0 40px; padding:20px 10px">' +
            '<div>' +
              '<img src="http://placehold.it/250x100&text=Another+Image" style="float:right; padding:10px"/>' +
              '<h2>No marker on map</h2>' +
              '<p>Aenean lacinia bibendum nulla sed consectetur. Fusce dapibus, tellus ac cursus commodo, tortor mauris condimentum nibh, ut fermentum massa justo sit amet risus. Aenean eu leo quam. Pellentesque ornare sem lacinia quam venenatis vestibulum.</p>' +
            '</div>' +
          '</div>'
  },

  // non-centered marker
  {
    marker: [36.573376, -115.674667],
    center: [35.573376, -121.674667],
    zoom: 6,
    html: '<div style="margin:0 40px; padding:20px 10px">' +
            '<div>' +
              '<h1>Non-centered marker</h1>' +
              '<p>Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Nulla vitae elit libero, a pharetra augue. Integer posuere erat a ante venenatis <a href="#">dapibus posuere</a> velit aliquet.</p><p>Cras justo odio, dapibus ac facilisis in, egestas eget quam. Sed posuere consectetur est at lobortis. Aenean lacinia bibendum nulla sed consectetur. Fusce dapibus, tellus ac cursus commodo, tortor mauris condimentum nibh, ut fermentum massa justo sit amet risus. Aenean eu leo quam. Pellentesque ornare sem lacinia quam venenatis vestibulum.</p>' +
            '</div>' +
          '</div>',
    popup: 'non-centered marker'
  },

  // clustered markers
  {
    marker: [40.072208, -105.508332],
    html: '<div style="margin:0 40px; padding:20px 10px">' +
            '<div>' +
              '<h1>Cluster marker 1</h1>' +
              '<p>Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Nulla vitae elit libero, a pharetra augue. Integer posuere erat a ante venenatis <a href="#">dapibus posuere</a> velit aliquet.</p><p>Cras justo odio, dapibus ac facilisis in, egestas eget quam. Sed posuere consectetur est at lobortis. Aenean lacinia bibendum nulla sed consectetur. Fusce dapibus, tellus ac cursus commodo, tortor mauris condimentum nibh, ut fermentum massa justo sit amet risus. Aenean eu leo quam. Pellentesque ornare sem lacinia quam venenatis vestibulum.</p>' +
            '</div>' +
          '</div>',
    popup: 'cluster marker #1'
  },
  {
    marker: [39.961376, -105.510831],
    html: '<div style="margin:0 40px; padding:20px 10px">' +
            '<div>' +
              '<h1>Cluster marker 2</h1>' +
              '<p>Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Nulla vitae elit libero, a pharetra augue. Integer posuere erat a ante venenatis <a href="#">dapibus posuere</a> velit aliquet.</p><p>Cras justo odio, dapibus ac facilisis in, egestas eget quam. Sed posuere consectetur est at lobortis. Aenean lacinia bibendum nulla sed consectetur. Fusce dapibus, tellus ac cursus commodo, tortor mauris condimentum nibh, ut fermentum massa justo sit amet risus. Aenean eu leo quam. Pellentesque ornare sem lacinia quam venenatis vestibulum.</p>' +
            '</div>' +
          '</div>',
    popup: 'cluster marker #2'
  },
  {
    marker: [39.737567, -104.984718],
    html: '<div style="margin:0 40px; padding:20px 10px">' +
            '<div>' +
              '<h1>Cluster marker 3</h1>' +
              '<p>Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Nulla vitae elit libero, a pharetra augue. Integer posuere erat a ante venenatis <a href="#">dapibus posuere</a> velit aliquet.</p><p>Cras justo odio, dapibus ac facilisis in, egestas eget quam. Sed posuere consectetur est at lobortis. Aenean lacinia bibendum nulla sed consectetur. Fusce dapibus, tellus ac cursus commodo, tortor mauris condimentum nibh, ut fermentum massa justo sit amet risus. Aenean eu leo quam. Pellentesque ornare sem lacinia quam venenatis vestibulum.</p>' +
            '</div>' +
          '</div>',
    popup: 'cluster marker #3'
  },
  {
    marker: [37.811941, -107.664506],
    html: '<div style="margin:0 40px; padding:20px 10px">' +
            '<div>' +
              '<h1>Cluster marker 4</h1>' +
              '<p>Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Nulla vitae elit libero, a pharetra augue. Integer posuere erat a ante venenatis <a href="#">dapibus posuere</a> velit aliquet.</p><p>Cras justo odio, dapibus ac facilisis in, egestas eget quam. Sed posuere consectetur est at lobortis. Aenean lacinia bibendum nulla sed consectetur. Fusce dapibus, tellus ac cursus commodo, tortor mauris condimentum nibh, ut fermentum massa justo sit amet risus. Aenean eu leo quam. Pellentesque ornare sem lacinia quam venenatis vestibulum.</p>' +
            '</div>' +
          '</div>',
    popup: 'cluster marker #4'
  },
  {
    marker: [40.377206, -105.521665],
    html: '<div style="margin:0 40px; padding:20px 10px">' +
            '<div>' +
              '<h1>Cluster marker 5</h1>' +
              '<p>Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Praesent commodo cursus magna, vel scelerisque nisl consectetur et. Cras justo odio, dapibus ac facilisis in, egestas eget quam. Nulla vitae elit libero, a pharetra augue. Integer posuere erat a ante venenatis <a href="#">dapibus posuere</a> velit aliquet.</p><p>Cras justo odio, dapibus ac facilisis in, egestas eget quam. Sed posuere consectetur est at lobortis. Aenean lacinia bibendum nulla sed consectetur. Fusce dapibus, tellus ac cursus commodo, tortor mauris condimentum nibh, ut fermentum massa justo sit amet risus. Aenean eu leo quam. Pellentesque ornare sem lacinia quam venenatis vestibulum.</p>' +
            '</div>' +
          '</div>',
    popup: 'cluster marker #5'
  },

  // empty slide
  {
    html: '<div style="margin:0 40px; padding:20px 10px">' +
            '<div>' +
              '<h2>The End</h2>' +
              '<p>Goodbye.</p>' +
            '</div>' +
          '</div>'
  }

];


slidemapper.js
/* ==========================================================
 * slidemapper.js v0.1.2
 * A jQuery plugin to create an animated slideshow tied to a Leaflet map.
 * http://github.com/cavis/slidemapper
 * ==========================================================
 * Copyright (c) 2012 Ryan Cavis
 * Licensed under http://en.wikipedia.org/wiki/MIT_License
 * ========================================================== */


/* ==========================================================
 * SlideMapper jquery extension
 *
 * Creates a slidemapper widget with an element.
 * ========================================================== */
(function($) {

  // default configuration options
  var defaultOptions = {

    // an initial set of slides to add to the map
    // (optional... you can always add slides later)
    slides: [],

    // the type of map to use. valid options are: cloudmade / stamen-toner /
    // stamen-terrain / stamen-watercolor / mapquest / mapquest-aerial
    mapType: 'mapquest',

    // if using cloudmade tiles, provide your apikey here
    apiKey:  null,

    // the default/fallback center and zoomlevel of the map
    center: [40.423, -98.7372],
    zoom: 4,

    // minimum and maximum allowed zoom levels
    minZoom: 2,
    maxZoom: 10,

    // enable key events (left-right arrow keys)
    keyEvents: true,

    // allow popups to close on the map
    closePopupOnClick: false,

    // height/layout settings
    mapPosition: 'bottom', //top|bottom
    mapHeight: 400,
    slideHeight: 220, //if autoHeight, this serves as the min slide height
    autoHeight: false,
    leafPile: false,

    // slideshow settings
    animateSpeed: 200,
    controlType: 'sides' // sides|top
  };


  // private vars, defined at the beginning of every call, saved at the end
  var DATA;
  var $THIS;


  // private methods
  function _slideOut($el, goLeft) {
    var end = goLeft ? -($el.width()) : $el.width();
    $el.animate({'margin-left': end}, DATA.options.animateSpeed, 'swing', function() { $el.removeClass('active').removeAttr('style'); });
  }
  function _slideIn($el, goLeft) {
    var start = goLeft ? $el.width() : -($el.width());
    $el.css('margin-left', start).addClass('active');
    $el.animate({'margin-left': 0}, DATA.options.animateSpeed, 'swing', function() { $el.removeAttr('style'); });
    _autoHeight($el, true);
  }
  function _refreshControls() {
    var left = $THIS.find('.ctrl-left'),
        right = $THIS.find('.ctrl-right'),
        count = $THIS.find('.ctrl-count');
    (DATA.index > 0) ? left.addClass('active') : left.removeClass('active');
    (DATA.index < DATA.items.length-1) ? right.addClass('active') : right.removeClass('active');
    count.html((DATA.index === null ? 0 : DATA.index+1) + ' of ' + DATA.items.length);
  }
  function _autoHeight($el, animate) {
    if (DATA.options.autoHeight) {
      var $show = $THIS.find('.smapp-show');
      if (!DATA.autoHeight) DATA.autoHeight = $show.height();
      var inner = $el.find('.slide-inner').height(), outer = $show.height();
      if (inner > outer) {
        animate ? $show.animate({'height': inner}, DATA.options.animateSpeed) : $show.height(inner);
      }
      else if (inner < outer) {
        var newH = Math.max(inner, DATA.autoHeight);
        animate ? $show.animate({'height': newH}, DATA.options.animateSpeed) : $show.height(newH);
      }
    }
  }
  function _onKeyPress(e) {
    if (DATA.index !== null) {
      if (e.keyCode == 37) $THIS.slideMapper('prev');
      if (e.keyCode == 39) $THIS.slideMapper('next');
    }
  }
  function _setTiles(tileType) {
    if (DATA.tileLayer) {
      DATA.map.removeLayer(DATA.tileLayer);
      DATA.tileLayer = false;
    }

    // set the new tile layer
    if (tileType == 'cloudmade') {
      if (!DATA.options.apiKey) $.error('apiKey required for cloudmade tiles');
      var tileOpts = {attribution: 'Map data &copy; <a href="http://openstreetmap.org">OpenStreetMap</a> contributors, <a href="http://creativecommons.org/licenses/by-sa/2.0/">CC-BY-SA</a>, Imagery © <a href="http://cloudmade.com">CloudMade</a>'};
      DATA.tileLayer = new L.TileLayer('http://{s}.tile.cloudmade.com/'+DATA.options.apiKey+'/997/256/{z}/{x}/{y}.png', tileOpts);
    }
    else if (tileType == 'stamen-toner') {
      if (!L.StamenTileLayer) $.error('did you forget to include tile.stamen.js?');
      DATA.tileLayer = new L.StamenTileLayer('toner');
    }
    else if (tileType == 'stamen-terrain') {
      if (!L.StamenTileLayer) $.error('did you forget to include tile.stamen.js?');
      DATA.tileLayer = new L.StamenTileLayer('terrain');
    }
    else if (tileType == 'stamen-watercolor') {
      if (!L.StamenTileLayer) $.error('did you forget to include tile.stamen.js?');
      DATA.tileLayer = new L.StamenTileLayer('watercolor');
    }
    else if (tileType == 'mapquest') {
      var tileOpts = {
        subdomains:  ['otile1', 'otile2', 'otile3', 'otile4'],
        attribution: 'Tiles Courtesy of <a href="http://www.mapquest.com/" target="_blank">MapQuest</a> <img src="http://developer.mapquest.com/content/osm/mq_logo.png">'
      };
      DATA.tileLayer = new L.TileLayer('http://{s}.mqcdn.com/tiles/1.0.0/osm/{z}/{x}/{y}.png', tileOpts);
    }
    else if (tileType == 'mapquest-aerial') {
      var tileOpts = {
        subdomains:  ['oatile1', 'oatile2', 'oatile3', 'oatile4'],
        attribution: 'Tiles Courtesy of <a href="http://www.mapquest.com/" target="_blank">MapQuest</a> <img src="http://developer.mapquest.com/content/osm/mq_logo.png">'
      };
      DATA.tileLayer = new L.TileLayer('http://{s}.mqcdn.com/tiles/1.0.0/sat/{z}/{x}/{y}.png', tileOpts);
    }
    else {
      $.error('invalid tile type: '+tileType);
    }
    DATA.map.addLayer(DATA.tileLayer);
  }
  function _makeHTML(opts) {
    var map, mapStyle = 'style="height:'+opts.mapHeight+'px"';
    if (opts.mapPosition === 'top') {
      map = '<div class="smapp-map top" '+mapStyle+'></div>';
    }
    else if (opts.mapPosition === 'bottom') {
      map = '<div class="smapp-map bottom" '+mapStyle+'></div>';
    }
    else {
      $.error('Invalid mapPosition: '+opts.mapPosition);
    }

    // slideshow gets either height or min-height
    var show = opts.autoHeight
      ? '<div class="smapp-show" style="min-height:'+opts.slideHeight+'px">'
      : '<div class="smapp-show" style="height:'+opts.slideHeight+'px">';
    show += '<div class="smapp-slides-ct"></div>';

    // controls
    if (opts.controlType === 'sides') {
      show += '<span class="ctrl-left ctrl-side">&lsaquo;</span>';
      show += '<span class="ctrl-right ctrl-side">&rsaquo;</span>';
    }
    else if (opts.controlType === 'top') {
      show += '<div class="ctrl-top">';
      show += '<span class="ctrl-left">&lsaquo;</span>';
      show += '<span class="ctrl-count">0 of 0</span>';
      show += '<span class="ctrl-right">&rsaquo;</span>';
      show += '</div>';
    }
    else {
      $.error('Invalid controlType: '+opts.controlType);
    }
    show += '</div>'; //end smapp-show

    // assemble in container
    var s = '<div class="smapp">';
    s += (opts.mapPosition === 'top') ? map : '';
    s += show;
    s += (opts.mapPosition === 'bottom') ? map : '';
    return s;
  }
  function _showPopup(item, panTo) {
    DATA.map.closePopup();
    if (item.marker && item.config.popup) {
      if (item.marker._leafpile) {
        item.marker.openPopup();
      }
      else {
        var popup = item.marker._popup.setLatLng(item.marker.getLatLng());
        DATA.map.openPopup(popup);
      }
    }

    // determine the center
    DATA.map.setView(item.center, item.zoom || DATA.map.getZoom(), panTo ? false : true);
  }
  function _getMarkerLayer() {
    var markerLayer;

    // create a leafpile, or a normal layer
    if (DATA.options.leafPile) {
      if (!L.LeafpileGroup) $.error('Did you forget to include leafpile.js?');
      var opts = DATA.options.leafPile === true ? {} : DATA.options.leafPile;
      markerLayer = new L.LeafpileGroup(opts);

      // listen to zooms/redraws
      markerLayer.on('redraw', function(e) {
        if (DATA.index !== null && DATA.items[DATA.index].config.popup) {
          DATA.items[DATA.index].marker.openPopup();
        }
      });
      markerLayer.on('click', function(e) {
        // slide the first marker in
        if (e.leafpile) {
          var idx = e.markers[0].index;
          if (idx != DATA.index) {
            if (DATA.index !== null) _slideOut(DATA.items[DATA.index].$slide, (idx > DATA.index));
            _slideIn(DATA.items[idx].$slide, (idx > DATA.index));
            DATA.index = idx;
            _refreshControls();
          }
        }

        // cancel if frozen
        if (e.zooming && DATA.frozen) {
          e.cancelZoom();
        }
      });
    }
    else {
      markerLayer = new L.LayerGroup();
    }
    return markerLayer;
  }


  // public methods
  var methods = {

    // initial setup
    init: function(passedOpts) {
      if (!DATA) {
        DATA = {};
        DATA.options = $.extend({}, defaultOptions, passedOpts);
        methods.options(passedOpts || {});
      }
    },

    // get or set map options
    options: function(passedOpts) {
      if (passedOpts === undefined) {
        return DATA.options;
      }
      else {
        DATA.options = $.extend({}, DATA.options, passedOpts);
        $THIS.empty();
        $THIS.append(_makeHTML(DATA.options));
        showEl = $('.smapp-show', $THIS)[0];
        mapEl  = $('.smapp-map',  $THIS)[0];

        // left/right listeners
        $(showEl)
          .on('click', '.ctrl-left', function() {$THIS.slideMapper('prev');})
          .on('click', '.ctrl-right', function() {$THIS.slideMapper('next');});
        methods.keyEvents(DATA.options.keyEvents);

        // initialize the map
        DATA.options.center = new L.LatLng(DATA.options.center[0], DATA.options.center[1]);
        DATA.map = new L.Map(mapEl, DATA.options);
        _setTiles(DATA.options.mapType);

        // create a layer for the markers
        DATA.markergroup = _getMarkerLayer();
        DATA.map.addLayer(DATA.markergroup);

        // setup initial items
        DATA.items = [];
        DATA.index = null;
        methods.add(DATA.options.slides);
      }
    },

    // set key events on or off
    keyEvents: function(turnOn) {
      $(document).unbind('keydown', _onKeyPress);
      if (turnOn) {
        $(document).keydown(_onKeyPress)
      }
    },

    // set map events on or off
    mapEvents: function(turnOn) {
      var props = ['dragging', 'touchZoom', 'doubleClickZoom', 'scrollWheelZoom', 'boxZoom'];
      var fn = turnOn ? 'enable' : 'disable';
      for (var i=0; i<props.length; i++) {
        var p = props[i];
        if (DATA.map[p] && typeof(DATA.map[p][fn]) === 'function') {
          DATA.map[p][fn]();
        }
      }
    },

    // prevent slide from changing
    freeze: function(makeFrozen) {
      DATA.frozen = makeFrozen;
      makeFrozen ? methods.keyEvents(false) : methods.keyEvents(true);
    },

    // get the number of items in the slideshow
    count: function() {
      return DATA.items.length;
    },

    // get the slide at an index (or current index if null)
    get: function(index) {
      index = (index === undefined || index === null) ? DATA.index : index;
      return (DATA.items[index] === undefined) ? false : DATA.items[index];
    },

    // append a slide (or slides) to the end of the show
    add: function(cfg) {
      var idx = (DATA.items && DATA.items.length) ? DATA.items.length : 0;
      return methods.insert(idx, cfg);
    },

    // insert a slide (or slides) into the show
    insert: function(index, cfg) {
      if ($.isArray(cfg)) {
        var ins = [];
        $.each(cfg, function(i, c) { ins.push(methods.insert.call(this, index+i, c)); });
        return ins;
      }
      if (!$.isNumeric(index) || index < 0 || index > DATA.items.length) {
        $.error('Invalid index "'+index+'" provided to insert');
      }
      var item = {index: index, config: cfg};

      // add marker to map
      if (cfg.marker) {

        // adds class to extend leaflet marker
        var LeafIcon = L.Icon.extend({
            options: {
                shadowUrl: '',
                iconSize:     [40, 40],
                shadowSize:   [0, 0],
                iconAnchor:   [22, 94],
                shadowAnchor: [0, 0],
                popupAnchor:  [-3, -76]
            }
        });

        var latlng = new L.LatLng(cfg.marker[0], cfg.marker[1]);

        // creates instance of map icon
        if (cfg.icon) {
            var mapIcon = new LeafIcon({iconUrl: cfg.icon});
            item.marker = new L.marker(latlng, {icon: mapIcon});
        } else {
            item.marker = new L.Marker(latlng);
        }

        if (cfg.popup) item.marker.bindPopup(cfg.popup);

        item.marker.index = index;
        item.marker.on('click', function(e) {
          var idx = e.target.index;
          if (!DATA.frozen && DATA.index != idx && $THIS.triggerHandler('move', [methods.get(idx), idx]) !== false) {
            if (DATA.index !== null) _slideOut(DATA.items[DATA.index].$slide, (idx > DATA.index));
            _slideIn(DATA.items[idx].$slide, (idx > DATA.index));
            DATA.index = idx;
            _refreshControls();
          }
        });
        DATA.markergroup.addLayer(item.marker);
      }

      // render to slide
      var html = '<div class="slide"><div class="slide-inner">'+(cfg.html || '')+'</div></div>';
      if (index == DATA.items.length) {
        item.$slide = $(html).appendTo($THIS.find('.smapp-slides-ct'));
      }
      else {
        item.$slide = $(html).insertBefore(DATA.items[index].$slide);
      }

      // figure out center and zoom (optional)
      item.zoom = item.marker ? cfg.zoom : DATA.options.zoom;
      item.center = item.marker ? item.marker.getLatLng() : DATA.options.center;
      if (cfg.center) {
        item.center = new L.LatLng(cfg.center[0], cfg.center[1]);
      }

      // splice into data items
      DATA.items.splice(index, 0, item);
      for (var i=index+1; i<DATA.items.length; i++) {
        DATA.items[i].index++;
        if (DATA.items[i].marker) DATA.items[i].marker.index++;
      }

      // show slide and refresh controls
      if (DATA.items.length == 1) {
        methods.move(0, false); //initial
      }
      else if (DATA.index === index) {
        DATA.index++; //follow the active slide
      }
      _refreshControls();
      return item;
    },

    // move a slide to a new position
    shuffle: function(index1, index2) {
      var fidx = (index2 === undefined || index2 === null) ? DATA.index : index1;
      var item = methods.get(fidx);
      if (!item) {
        $.error('Invalid from index "'+fidx+'" provided to remove');
      }

      var tidx = (index2 === undefined || index2 === null) ? index1 : index2;
      if (tidx == 'first') tidx = 0;
      if (tidx == 'last') tidx = DATA.items.length-1;
      if (!$.isNumeric(tidx) || tidx < 0 || tidx > DATA.items.length-1) {
        $.error('Invalid to index "'+tidx+'" provided to insert');
      }
      if (fidx == tidx) return; //nothing to do

      // splice out of items
      DATA.items.splice(fidx, 1);
      for (var i=fidx; i<DATA.items.length; i++) {
        DATA.items[i].index--;
        if (DATA.items[i].marker) DATA.items[i].marker.index--;
      }

      // move the slide
      if (tidx == DATA.items.length) {
        item.$slide.appendTo($THIS.find('.smapp-slides-ct'));
      }
      else {
        item.$slide.insertBefore(DATA.items[tidx].$slide);
      }

      // splice back into items and update index
      item.index = tidx;
      if (item.marker) item.marker.index = tidx;
      DATA.items.splice(tidx, 0, item);
      for (var i=tidx+1; i<DATA.items.length; i++) {
        DATA.items[i].index++;
        if (DATA.items[i].marker) DATA.items[i].marker.index++;
      }

      // follow me
      if (DATA.index == fidx) {
        DATA.index = tidx;
      }
      _refreshControls();
      return item;
    },

    // remove a single slide/marker
    remove: function(index) {
      var item = methods.get(index);
      if (!item) {
        $.error('Invalid index "'+index+'" provided to remove');
      }
      if (DATA.items.length == 1) {
        return methods.removeAll(); //just nuke it
      }

      // remove from map, slide, and items
      if (item.marker) {
        DATA.markergroup.removeLayer(item.marker);
      }
      item.$slide.remove();
      DATA.items.splice(item.index, 1);
      for (var i=item.index; i<DATA.items.length; i++) {
        DATA.items[i].index--;
        if (DATA.items[i].marker) DATA.items[i].marker.index--;
      }

      // show slide and refresh controls
      if (DATA.index === item.index) {
        var flipTo = Math.min(item.index, DATA.items.length-1);
        _slideIn(DATA.items[flipTo].$slide, (flipTo >= DATA.index));
        _showPopup(DATA.items[flipTo], true);
        DATA.index = flipTo;
      }
      _refreshControls();
    },

    // fast remove all slides/markers
    removeAll: function() {
      DATA.map.removeLayer(DATA.markergroup);
      $THIS.find('.smapp-slides-ct').empty();

      // re-init the blank map
      DATA.markergroup = _getMarkerLayer();
      DATA.map.addLayer(DATA.markergroup);
      DATA.items = [];
      DATA.index = null;
      _refreshControls();

      // re-center the blank map
      DATA.map.setView(DATA.options.center, DATA.options.zoom);
    },

    // move to a different marker
    move: function(index, animate) {
      if (DATA.frozen) return;
      if (!methods.get(index)) return;
      if ($THIS.triggerHandler('move', [methods.get(index), index]) === false) return;

      // slide out the old, in the new preview
      if (animate) {
        if (DATA.index !== null) _slideOut(DATA.items[DATA.index].$slide, (index > DATA.index));
        _slideIn(DATA.items[index].$slide, (index > DATA.index));
      }
      else {
        if (DATA.index !== null) DATA.items[DATA.index].$slide.removeClass('active');
        DATA.items[index].$slide.addClass('active');
        _autoHeight(DATA.items[0].$slide, false);
      }

      // open new popup and update stored index
      _showPopup(DATA.items[index], animate);
      DATA.index = index;
      _refreshControls();
    },

    // next!
    next: function() {
      methods.move(DATA.index === null ? 0 : DATA.index + 1, true);
    },

    // previous!
    prev: function() {
      methods.move(DATA.index === null ? 0 : DATA.index - 1, true);
    }

  };

  // attach jquery namespace
  $.fn.slideMapper = function(method) {
    if (this.length > 1) {
      $.error('SlideMapper currently only supports 1 map per page');
    }
    else if (method && typeof method !== 'object' && !methods[method]) {
      $.error('Method '+method+' does not exist on jQuery.slideMapper');
    }
    else {
      var ret = null;

      // call for each element
      for (var i=0; i<this.length; i++) {

        // setup private vars
        $THIS = $(this[i]);
        DATA  = $THIS.data('slideMapper');

        // call init if no method given
        if (methods[method]) {
          if (!DATA) $.error('Method '+method+' called on uninitialized element');
          else ret = methods[method].apply(this[i], Array.prototype.slice.call(arguments, 1));
        }
        else {
          methods.init.apply(this[i], arguments);
        }

        // save data changes
        $THIS.data('slideMapper', DATA);
      }
      return (ret === null ? this : ret);
    }
  };

}) (jQuery);
